# Hypothesis formulations on the attrition analysis

Attach the dataset and mentioned problem statement below,

* A sizable department has to be maintained, for the purposes of recruiting new talent
* More often than not, the new employees have to be trained for the job and/or given time to acclimatize themselves to the company
* Hence, the management has contracted an HR analytics firm to understand what factors they should focus on, in order to curb attrition.
* In other words, they want to know what changes they should make to their workplace, in order to get most of their employees to stay.
* Also, they want to know which of these variables is most important and needs to be addressed right away.

Load the data

In [1]:
import pandas as pd

data = pd.read_csv("general_data.csv")

df = pd.DataFrame(data)

check the data

In [2]:
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

check the columns

In [3]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

Null value check

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

Analyse "TotalWorkingYears" data descripency

In [11]:
df[df["TotalWorkingYears"].isnull()]

Age Attrition     BusinessTravel              Department  \
23     42        No      Travel_Rarely  Research & Development   
137    43        No  Travel_Frequently                   Sales   
308    47       Yes  Travel_Frequently  Research & Development   
574    37        No      Travel_Rarely  Research & Development   
1517   36        No         Non-Travel  Research & Development   
2367   39        No      Travel_Rarely                   Sales   
3120   35        No      Travel_Rarely  Research & Development   
3818   28       Yes      Travel_Rarely         Human Resources   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
23                   4          4  Life Sciences              1          24   
137                 14          3  Life Sciences              1         138   
308                  4          3  Life Sciences              1         309   
574                 23          3  Life Sciences              1         575   
1517                 5          4  Life Sciences              1        1518   
2367                 2          4  Life Sciences              1        2368   
3120                12          2        Medical              1        3121   
3818                 1          3        Medical              1        3819   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
23      Male  ...                 1.0      Y                14              8   
137   Female  ...                 7.0      Y                14              8   
308     Male  ...                 4.0      Y                13              8   
574     Male  ...                 7.0      Y                16              8   
1517    Male  ...                 7.0      Y                21              8   
2367    Male  ...                 0.0      Y                17              8   
3120  Female  ...                 4.0      Y                23              8   
3818  Female  ...                 1.0      Y                16              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
23                   0               NaN                      4   
137                  2               NaN                      3   
308                  1               NaN                      2   
574                  1               NaN                      2   
1517                 0               NaN                      4   
2367                 0               NaN                      5   
3120                 1               NaN                      2   
3818                 1               NaN                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
23                20                       11                     6  
137                3                        1                     2  
308                5                        1                     3  
574                6                        1                     3  
1517               9                        0                     8  
2367              10                        0                     7  
3120               8                        0                     0  
3818               1                        0                     0  
4409              21                        3                     9  

[9 rows x 24 columns]

In [12]:
df[df["NumCompaniesWorked"].isnull()]

Age Attrition     BusinessTravel              Department  \
115    27        No      Travel_Rarely                   Sales   
210    52        No      Travel_Rarely                   Sales   
343    40        No  Travel_Frequently                   Sales   
476    22        No      Travel_Rarely  Research & Development   
647    29        No      Travel_Rarely  Research & Development   
799    29       Yes  Travel_Frequently  Research & Development   
932    32        No      Travel_Rarely  Research & Development   
1103   26        No  Travel_Frequently                   Sales   
1312   37       Yes      Travel_Rarely         Human Resources   
1521   31        No      Travel_Rarely  Research & Development   
1711   33        No      Travel_Rarely                   Sales   
1996   38        No      Travel_Rarely  Research & Development   
2365   18       Yes  Travel_Frequently  Research & Development   
2696   28       Yes      Travel_Rarely  Research & Development   
3063   30        No      Travel_Rarely  Research & Development   
3533   56        No      Travel_Rarely  Research & Development   
3910   41        No         Non-Travel                   Sales   
4226   36        No      Travel_Rarely  Research & Development   
4395   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
115                  2          3     Life Sciences              1   
210                  2          1           Medical              1   
343                  9          3         Marketing              1   
476                  8          3             Other              1   
647                  3          3  Technical Degree              1   
799                 28          3           Medical              1   
932                  5          4  Technical Degree              1   
1103                16          2     Life Sciences              1   
1312                 9          3   Human Resources              1   
1521                20          1  Technical Degree              1   
1711                13          4         Marketing              1   
1996                 5          3  Technical Degree              1   
2365                 3          5           Medical              1   
2696                 1          4     Life Sciences              1   
3063                 3          3     Life Sciences              1   
3533                 3          2             Other              1   
3910                 2          2           Medical              1   
4226                 2          3     Life Sciences              1   
4395                 2          3     Life Sciences              1   

      EmployeeID  Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  \
115          116    Male  ...                 NaN      Y                12   
210          211    Male  ...                 NaN      Y                11   
343          344    Male  ...                 NaN      Y                14   
476          477    Male  ...                 NaN      Y                21   
647          648  Female  ...                 NaN      Y                16   
799          800  Female  ...                 NaN      Y                17   
932          933    Male  ...                 NaN      Y                13   
1103        1104    Male  ...                 NaN      Y                14   
1312        1313    Male  ...                 NaN      Y                17   
1521        1522  Female  ...                 NaN      Y                13   
1711        1712  Female  ...                 NaN      Y                18   
1996        1997    Male  ...                 NaN      Y                11   
2365        2366    Male  ...                 NaN      Y                12   
2696        2697    Male  ...                 NaN      Y                13   
3063        3064  Female  ...                 NaN      Y                13   
3533        3534    Male  ...       

Data descripency observed on NumCompaniesWorked and TotalWorkingYears are because of the NaN value. Replacing the NaN values with 0

In [15]:
df['NumCompaniesWorked'] = df['NumCompaniesWorked'].fillna(0)
df['TotalWorkingYears'] = df['TotalWorkingYears'].fillna(0)

Checking the corrected data

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

Checking the general statistics on the data when attrition is true

In [24]:
df[df.Attrition == "Yes"].describe()

Age  DistanceFromHome   Education  EmployeeCount   EmployeeID  \
count  711.000000        711.000000  711.000000          711.0   711.000000   
mean    33.607595          9.012658    2.877637            1.0  2191.767932   
std      9.675693          7.772368    1.014233            0.0  1270.136654   
min     18.000000          1.000000    1.000000            1.0     2.000000   
25%     28.000000          2.000000    2.000000            1.0  1072.000000   
50%     32.000000          7.000000    3.000000            1.0  2165.000000   
75%     39.000000         15.000000    4.000000            1.0  3341.000000   
max     58.000000         29.000000    5.000000            1.0  4403.000000   

         JobLevel  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
count  711.000000     711.000000          711.000000         711.000000   
mean     2.037975   61682.616034            2.919831          15.481013   
std      1.057485   44792.067695            2.680243           3.775289   
min      1.000000   10090.000000            0.000000          11.000000   
25%      1.000000   28440.000000            1.000000          12.000000   
50%      2.000000   49080.000000            1.000000          14.000000   
75%      2.000000   71040.000000            5.000000          18.000000   
max      5.000000  198590.000000            9.000000          25.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count          711.0        711.000000         711.000000   
mean             8.0          0.780591           8.232068   
std              0.0          0.858899           7.167288   
min              8.0          0.000000           0.000000   
25%              8.0          0.000000           3.000000   
50%              8.0          1.000000           7.000000   
75%              8.0          1.000000          10.000000   
max              8.0          3.000000          40.000000   

       TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count             711.000000      711.000000               711.000000   
mean                2.654008        5.130802                 1.945148   
std                 1.154834        5.941598                 3.148633   
min                 0.000000        0.000000                 0.000000   
25%                 2.000000        1.000000                 0.000000   
50%                 3.000000        3.000000                 1.000000   
75%                 3.000000        7.000000                 2.000000   
max                 6.000000       40.000000                15.000000   

       YearsWithCurrManager  
count            711.000000  
mean               2.852321  
std                3.138918  
min                0.000000  
25%                0.000000  
50%                2.000000  
75%                5.000000  
max               14.000000

Hypothesis formulations

* Attrition is caused due to employees below 33 years of age
* Attrition is caused due to distance from home more than 9
* Attrition is caused due to Monthly salary less than 61682
* Attrition is caused due to number of companies woked is more than or equal to 2
* Attrition is caused due to salary hike less than 2.9
* Attrition is caused due to total working years less than 8
* Attrition is caused due to tatal years at company is less than 5
* Attrition is caused due to exceeding last promotion more than 2 
* Attrition is caused due to deviation in salary hike more than 2
* Attrition is caused due to stock option level less than 0.78